### PR TITLE
Feature/point cloud cs

### PR DESCRIPTION
**Pull request by kukash**
Created a point cloud generator.
The pointcloud generation system creates a point cloud based on point_cloud components using a openCl compute shader for sampling and the particle system for rendering.
Unbenannt

### CODE REVIEW

File: applications/sandbox/assets/kernels/pointRasterizer.cl [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+//constant values
+//exit condition for poission sampling
+const int K=30;
+const uint RAND_MAX= 255;
+//predefined values
+#define maxPointsPerTri 300
+#define radius 0.33f
+#define cellSize (radius)/1.41421356237f
+#define depth (int)ceil(1/ cellSize)
+#define girdDimension  depth *depth
```
Comment (Algo-ryth-mix) [inline applications/sandbox/assets/kernels/pointRasterizer.cl]: typo ?

File: applications/sandbox/assets/kernels/pointRasterizer.cl [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+float RandomValue()
+{
+   uint randInt =Rand();
+
+   return randInt / (float) RAND_MAX;
+
+}
+
+//includes upper bound
+uint randomUpperRange(uint upper)
```
Comment (Algo-ryth-mix) [inline applications/sandbox/assets/kernels/pointRasterizer.cl]: Upper Case ?

**Review by tuhri444 — APPROVED** [review]

File: applications/sandbox/systems/PointCloudTestSystem.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
+#pragma once
+#include<core/core.hpp>
+#include <core/math/math.hpp>
+#include <core/logging/logging.hpp>
+#include <application/application.hpp>
```
Comment (Algo-ryth-mix) [inline applications/sandbox/systems/PointCloudTestSystem.hpp]: file should be `pointcloudtestsystem.hpp`

File: cl_error_patch.diff [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
++
++            size_t length;
++            char buffer[8192];
++
++            clGetProgramBuildInfo(m_program,device,CL_PROGRAM_BUILD_LOG,sizeof(buffer),buffer,&length);
++            buffer[length] = NULL;
++            log::warn("BUILD LOG:\n{}",buffer);
+         }
+ 
+         /*}*/
```
Comment (Algo-ryth-mix) [inline cl_error_patch.diff]: this file should not exist

File: include/core/common/result.hpp [hunk 5]
```diff
@@ -0,0 +1,457 @@
+#pragma once
+/**
+ * Copyright 2020 Raphael Baier, The Args Team
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files
```
Comment (Algo-ryth-mix) [inline include/core/common/result.hpp]: I am not fine with this

File: include/core/compute/high_level/function.hpp [hunk 6]
```diff
@@ -0,0 +1,326 @@
+#pragma once
+#include <type_traits>
+#include <variant>
+#include <tuple>
+#include <array>
```
Comment (Algo-ryth-mix) [inline include/core/compute/high_level/function.hpp]: same for this one (include files should not be commited)

File: legion/engine/core/compute/high_level/function.hpp [hunk 7]
```diff
@@ -21,35 +22,44 @@ namespace legion::core::compute {
...
 
     struct karg
     {
-        template <class T>
-        karg(T& v, const std::string& n = "") : container(&v,sizeof(T)), name(n) {}
+
+        karg(invalid_karg_type)
+        {
+
+        }
```
Comment (Algo-ryth-mix) [inline legion/engine/core/compute/high_level/function.hpp]: same line please

File: legion/engine/core/compute/high_level/function.hpp [hunk 8]
```diff
@@ -21,35 +22,44 @@ namespace legion::core::compute {
...
+        template <class T, std::enable_if_t<!std::is_same_v<std::remove_reference_t<T>, karg>, int > = 0>
+        karg(T& v, const std::string& n = "") : container(&v, sizeof(T)), name(n)
+        {
+            log::debug("hello");
+            log::debug(name);
```
Comment (Algo-ryth-mix) [inline legion/engine/core/compute/high_level/function.hpp]: remove debug info please

File: legion/engine/core/compute/high_level/function.hpp [hunk 9]
```diff
@@ -162,15 +172,41 @@ namespace legion::core::compute {
...
+            m_program = std::move(other.m_program);
+            m_kernel = std::move(other.m_kernel);
+            m_locals = std::move(other.m_locals);
+            return *this;
+        }
         /**
-         * @brief Sets the program from which to create the wrapped kernel.
-         */
+          * @brief Sets the program from which to create the wrapped kernel.
+          */
```
Comment (Algo-ryth-mix) [inline legion/engine/core/compute/high_level/function.hpp:197]: Please fix this indet

File: legion/engine/rendering/components/point_cloud.hpp [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            if (!m_mesh) return 0;
+            auto mesh = m_mesh.get();
+            auto indices = mesh.second.indices;
+            uint triangleCount = indices.size() / 3;
+            return maxPoints / triangleCount;
+        }
+    };
+
+
+
```
Comment (Algo-ryth-mix) [inline legion/engine/rendering/components/point_cloud.hpp]: I am judging your whitespace

File: legion/engine/rendering/components/point_cloud.hpp [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    struct point_cloud
+    {
+    public:
+
+        /**@brief Takes in a mesh handle, transform and a material, optionally it takes in max points and radius.
+         * Mesh is used to sample the points for the point cloud.
+         * Transform is the position of the object / point cloud.
+         * Material is the material used by the particle system that renders the points.
+         * Max points is used to calculate the samples for each triangle of the mesh,
+         * Radius defines point scale.
```
Comment (Algo-ryth-mix) [inline legion/engine/rendering/components/point_cloud.hpp]: at this point you might as well use `@param`

File: legion/engine/rendering/systems/pointcloudgeneration.hpp [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+
+namespace legion::rendering
+{
+
+    /**@class PointCloudGeneration
+     * @brief A system that iterates all queried entities containing point_cloud and generates a particle system for them.
+     */
+    class PointCloudGeneration : public System<PointCloudGeneration>
+
```
Comment (Algo-ryth-mix) [inline legion/engine/rendering/systems/pointcloudgeneration.hpp]: just no on this newl

**Review by Algo-ryth-mix — CHANGES_REQUESTED** [review]
Also please do not commit `*.lib` files they eat up a lot of lfs bandwith

File: applications/sandbox/assets/kernels/pointRasterizer.cl (hunk 2 above)
Comment (kukash) [reply applications/sandbox/assets/kernels/pointRasterizer.cl]: more detail pls

File: include/core/common/result.hpp (hunk 5 above)
Comment (kukash) [reply include/core/common/result.hpp]: what is going on here ?

Comment (Algo-ryth-mix) [reply include/core/common/result.hpp]: this is in the include folder, (aka should not be comitted)

**Review by Algo-ryth-mix — CHANGES_REQUESTED** [review]
some of the previous points still stand (include folder, excessive whitespace)

**Review by Algo-ryth-mix — APPROVED** [review]

**Pull request merged**